File: hard-hierarchy/memory-hierarchy.py
"""
Módulo: hierarquia_memoria.py
-----------------------------
Demonstra o conceito de hierarquia de memória conforme discutido
no livro "Fundamentos de Sistemas Operacionais" (Silberschatz).

Níveis simulados:
- Registradores: acessos ultrarrápidos (variáveis locais).
- Cache: arrays pequenos que cabem em cache L1/L2.
- RAM: arrays grandes, dependem da memória principal.

Objetivo:
Comparar os tempos médios de acesso em cada nível e evidenciar
as diferenças de desempenho entre registradores, cache e RAM.
"""

import time
import numpy as np


class MemoryHierarchySimulator:
    """
    Simula o acesso em diferentes níveis de memória:
    - Registradores (variáveis locais rápidas)
    - Cache (arrays pequenos, cabem na cache L1/L2)
    - RAM (arrays grandes, dependem de memória principal)
    """
    def __init__(self):
        self.register = 42
        self.cache = np.arange(10**4)   # tamanho pequeno (simula cache)
        self.ram = np.arange(10**7)     # tamanho grande (simula RAM)

    def _measure_access(self, operation):
        """
        Mede o tempo médio de execução de uma operação (5 repetições).
        :param operation: função que simula o acesso.
        :return: tempo médio em segundos.
        """
        times = []
        for _ in range(5):
            start = time.perf_counter()
            operation()
            end = time.perf_counter()
            times.append(end - start)
        return sum(times) / len(times)

    def access_register(self):
        """Simula acesso a registradores (variáveis locais)."""
        return self._measure_access(lambda: self.register * 2)

    def access_cache(self):
        """Simula acesso a cache (array pequeno)."""
        return self._measure_access(lambda: np.sum(self.cache))

    def access_ram(self):
        """Simula acesso à RAM (array grande)."""
        return self._measure_access(lambda: np.sum(self.ram))


if __name__ == "__main__":
    sim = MemoryHierarchySimulator()

    print("\n=== Simulação: Hierarquia de Memória ===")
    print("Medindo tempos médios de acesso (5 execuções cada)\n")

    reg_time = sim.access_register()
    cache_time = sim.access_cache()
    ram_time = sim.access_ram()

    # saída organizada em forma de tabela
    print(f"{'Nível de Memória':<20} | {'Descrição':<35} | {'Tempo Médio'}")
    print("-"*75)
    print(f"{'Registrador':<20} | {'Variável local (ultrarrápido)':<35} | {reg_time:.8f}s")
    print(f"{'Cache':<20}       | {'Array pequeno (cache L1/L2)':<35} | {cache_time:.6f}s")
    print(f"{'RAM':<20}         | {'Array grande (memória principal)':<35} | {ram_time:.4f}s")

    print("\nObservação:")
    print("- Registradores são centenas de vezes mais rápidos que RAM.")
    print("- Cache atua como intermediária, reduzindo latência.")
    print("- Isso demonstra por que a hierarquia de memória é essencial.\n")

File: hard-hierarchy/memoryHierarchy.cpp
#include <iostream>
#include <locale>
#include <codecvt>
#include <vector>
#include <cstdlib>
#include <cstdint>
#include <x86intrin.h>   // para __rdtsc (funciona em CPUs Intel/AMD com suporte a TSC)

// Função auxiliar para medir tempo em ciclos de CPU
inline uint64_t rdtsc() {
    return __rdtsc();
}

template<typename Func>
uint64_t measure(Func func, int iterations = 5) {
    uint64_t total = 0;
    for (int i = 0; i < iterations; i++) {
        uint64_t start = rdtsc();
        func();
        uint64_t end = rdtsc();
        total += (end - start);
    }
    return total / iterations;
}

int main() {
    // --- Simulação de diferentes níveis ---
    int reg = 42;                          // registrador
    std::vector<int> cache(10'000, 1);     // ~40 KB (cabe em cache L1/L2)
    std::vector<int> ram(50'000'000, 1);   // ~200 MB (força uso da RAM)

    // Medição
    auto reg_time = measure([&]() {
        int x = reg * 2;
    });

    auto cache_time = measure([&]() {
        long long sum = 0;
        for (int i = 0; i < cache.size(); i++)
            sum += cache[i];
    });

    auto ram_time = measure([&]() {
        long long sum = 0;
        for (int i = 0; i < ram.size(); i += 64)  // salto de 64 bytes (tamanho típico de linha de cache)
            sum += ram[i];
    });

    // Saída
    std::cout << "\n=== Medicao real de hierarquia de memoria (ciclos de CPU) ===\n\n";
    std::cout << "Nivel         | Descricao                         | Tempo MeSdio (ciclos)\n";
    std::cout << "-----------------------------------------------------------------\n";
    std::cout << "Registrador   | Variavel local (ultrarrapido)     | " << reg_time << "\n";
    std::cout << "Cache (L1/L2) | Array pequeno (~40KB)             | " << cache_time << "\n";
    std::cout << "RAM           | Array grande (~200MB)             | " << ram_time << "\n";

    std::cout << "\nObservacao:\n";
    std::cout << "- Registradores sao praticamente instantaneos.\n";
    std::cout << "- Cache L1/L2 atende rapidamente arrays pequenos.\n";
    std::cout << "- RAM tem latencia centenas de vezes maior.\n";
}
